### PR TITLE
Add post text to notification body

### PR DESCRIPTION
**Pull request by T-Damer**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -19,15 +19,19 @@
     "@hapi/boom": "^10.0.1",
     "@koa/cors": "^4.0.0",
     "@parse/node-apn": "^5.1.3",
+    "@types/node-fetch": "^2.6.2",
```
Comment (T-Damer) [inline package.json:22]: Types should be in `devDependencies {...}`, use `yarn install -D ...` to put stuff in there

File: src/helpers/api/validatePost.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+const checkStringData = (data?: string) => {
+  return typeof data === 'string' && data.length > 0
+}
+
+export default function (item: PostTexts): item is PostTexts {
+  const validText = 'text' in item && checkStringData(item.text)
+  const validExtraText =
+    'extraText' in item ? typeof item.extraText === 'string' : true
+
+  return validText && validExtraText
```
Comment (T-Damer) [inline src/helpers/api/validatePost.ts:12]: If somebody will comment, the post will be invalid?

File: src/helpers/generateRandomName.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { Config, uniqueNamesGenerator } from 'unique-names-generator'
+import {
+  animalDictionary,
+  colorDictionary,
+} from '@/helpers/data/NameDictionary'
+
+export default function (address: string) {
+  const customConfig: Config = {
```
Comment (T-Damer) [inline src/helpers/generateRandomName.ts:8]: You can put this object inside the function to get type annotations, so you don't need to defined `: Config`, which saves a little space

File: src/helpers/getIpfsMetadata.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  const cid = structToCid(metadata)
+
+  const response = await fetchWithTimeout(`${IPFS_GATEWAY}/${cid}`, {
+    headers,
+    timeout: 15000,
+  })
+
+  const data = (await response.json()) as ResponseData
+
+  cidToData.set(metadata, data)
```
Comment (T-Damer) [inline src/helpers/getIpfsMetadata.ts:24]: Why do you need this, if you don't `export/return cidToData` and don't use it :thinking:

File: src/helpers/api/index.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import { PostTexts } from '@/models/Post'
+import CID from '@/models/CID'
+import env from '@/helpers/env'
+import fetchWithTimeout from '@/helpers/fetchWithTimeout'
+import structToCid from '@/helpers/structToCid'
+import validatePost from '@/helpers/api/validatePost'
+
+const cidToData = new Map<CID, unknown>([])
```
Comment (T-Damer) [inline src/helpers/api/index.ts:8]: Why do you need this, if you don't `export/return cidToData` and don't use it :thinking:

File: src/helpers/sendGoogleNotification.ts [hunk 6]
```diff
@@ -3,23 +3,25 @@ import firebase from '@/helpers/firebase'
...
+function createNotification(title?: string, body?: string) {
   const baseNotification = {
     data: {
       type: 'newPost',
     },
   } as MessagingPayload
   if (title) {
     baseNotification.notification = {
-      title,
+      title: title,
```
Comment (T-Damer) [inline src/helpers/sendGoogleNotification.ts:14]: Don't duplicate `title: title`, you can use `title,` as shortcut, it automatically does `title: title`

File: src/helpers/sendAppleNotification.ts [hunk 7]
```diff
@@ -2,23 +2,37 @@ import { Notification } from '@parse/node-apn'
...
+function createNotification(
+  notificationTitle?: string,
+  notificationBody?: string
+) {
   const notification = new Notification()
   notification.topic = env.BUNDLE_ID
   notification.contentAvailable = true
   notification.priority = 5
   notification.pushType = 'background'
   if (notificationTitle) {
```
Comment (T-Damer) [inline src/helpers/sendAppleNotification.ts:14]: Should we also check for `body` here? What if there will be only title, will it be okay without body?

File: src/helpers/runNotificationListener.ts [hunk 8]
```diff
@@ -10,22 +14,33 @@ const rootFeeds = {
...
+    allTokens.forEach(async ({ token }) => {
+      try {
+        // APN token
+        if (apnRegex.test(token)) {
+          await sendAppleNotification(token, title, text)
+          console.log(text)
+        } else {
+          // FCM token
+          await sendGoogleNotification(token, title, text)
+          console.log('FCM token: ', token)
```
Comment (T-Damer) [inline src/helpers/runNotificationListener.ts:39]: Remove the `log`

File: src/helpers/runNotificationListener.ts [hunk 9]
```diff
@@ -10,22 +14,33 @@ const rootFeeds = {
...
+    const title = rootFeeds[feedId.toNumber()]
+      ? `@${nickname} posted at ${rootFeeds[feedId.toNumber()]}`
+      : undefined
+    const allTokens = await TokenModel.find()
+    allTokens.forEach(async ({ token }) => {
+      try {
+        // APN token
+        if (apnRegex.test(token)) {
+          await sendAppleNotification(token, title, text)
+          console.log(text)
```
Comment (T-Damer) [inline src/helpers/runNotificationListener.ts:35]: Remove the `log`

File: src/helpers/runNotificationListener.ts [hunk 10]
```diff
@@ -10,22 +14,33 @@ const rootFeeds = {
...
+obssContract.on(
+  'FeedPostAdded',
+  async (
+    feedId,
+    postId,
+    [author, metadata]: [author: string, metadata: CID]
+  ) => {
+    const { text } = await getPostByStruct(metadata)
+    const nickname = generateRandomName(author)
+    const title = rootFeeds[feedId.toNumber()]
```
Comment (T-Damer) [inline src/helpers/runNotificationListener.ts:26]: Extract `feedId.toNumber()` into separate `const`

File: src/helpers/generateRandomName.ts (hunk 3 above)
Comment (avr6ude) [reply src/helpers/generateRandomName.ts:8]: This is how it's implemented by the library authors, no need to save space.

File: src/helpers/sendGoogleNotification.ts [hunk 11]
```diff
@@ -3,23 +3,25 @@ import firebase from '@/helpers/firebase'
...
 const messaging = getMessaging(firebase)
 
-function createNotification(title?: string) {
+function createNotification(title?: string, body?: string) {
   const baseNotification = {
     data: {
       type: 'newPost',
     },
   } as MessagingPayload
   if (title) {
```
[the remark this answers is not in the document]
Comment (avr6ude) [reply src/helpers/sendGoogleNotification.ts:12]: No need. You can't send a post/comment without header. Body of notification is post header/comment.

File: src/helpers/runNotificationListener.ts [hunk 12]
```diff
@@ -10,22 +14,35 @@ const rootFeeds = {
...
+obssContract.on(
+  'FeedPostAdded',
+  async (
+    feedId,
+    postId,
+    [author, metadata]: [author: string, metadata: CID]
+  ) => {
+    const { text } = await getPostByStruct(metadata)
+    const nickname = generateRandomName(author)
+    const feed = feedId.toNumber()
```
Comment (T-Damer) [inline src/helpers/runNotificationListener.ts:26]: Rename it to `numberFeedId`

File: src/helpers/runNotificationListener.ts [hunk 13]
```diff
@@ -10,22 +14,35 @@ const rootFeeds = {
...
+
+    const title = rootFeeds[feed]
+      ? `@${nickname} posted at ${rootFeeds[feedId.toNumber()]}`
```
Comment (T-Damer) [inline src/helpers/runNotificationListener.ts:29]: Use `feed` const, don't convert `toNumber()` again

File: src/helpers/runNotificationListener.ts [hunk 14]
```diff
@@ -10,22 +14,35 @@ const rootFeeds = {
...
+      : undefined
+
+    const allTokens = await TokenModel.find()
+
+    allTokens.forEach(async ({ token }) => {
+      try {
+        // APN token
+        if (apnRegex.test(token)) {
+          await sendAppleNotification(token, title, text)
+        } else {
```
Comment (T-Damer) [inline src/helpers/runNotificationListener.ts:39]: Hm, should we also test google token :thinking:

File: src/helpers/api/index.ts [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+export async function getMetadataFromIpfs<ResponseData>(metadata: CID) {
+  const cid = structToCid(metadata)
+
+  const response = await fetchWithTimeout(`${IPFS_GATEWAY}/${cid}`, {
+    headers,
+    timeout: 15000,
+  })
+
+  const data = (await response.json()) as ResponseData
```
Comment (T-Damer) [inline src/helpers/api/index.ts:22]: Don't create `const data`, do:
```TS
return (await response.json()) as ResponseData
```

Comment (T-Damer) [reply src/helpers/api/index.ts:22]: Or specify return type where you declare the function

**Pull request closed without merging**